crates/ott-common/src/discovery/dns.rs

```diff
@@ -21,6 +25,44 @@ pub struct DnsDiscoveryConfig {
     pub polling_interval: Option<Duration>,
 }
 
+trait FromValue {
+    fn from_value(value: serde_json::Value) -> Result<DnsDiscoveryConfig, serde_json::Error>;
+}
+
+impl FromValue for DnsDiscoveryConfig {
+    fn from_value(mut value: serde_json::Value) -> Result<DnsDiscoveryConfig, serde_json::Error> {
+        if value.get("dns_server").is_none() {
+            let mut kv = serde_json::Map::new();
+            kv.insert(
+                "dns_server".to_string(),
+                serde_json::Value::from_str("").unwrap(),
+            );
+            let _value = value.as_object_mut().insert(&mut kv);
+        }
+
+        serde_json::from_value(value)
+    }
+}
+
+fn deserialize_dns_server<'de, D>(deserializer: D) -> Result<Option<SocketAddr>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let buf = String::deserialize(deserializer)?;
+
+    if buf.is_empty() {
+        return Ok(None);
+    }
+
+    match IpAddr::from_str(&buf) {
+        Ok(ip) => Ok(Some(SocketAddr::new(ip, 53))),
+        Err(_) => match SocketAddr::from_str(&buf) {
+            Ok(socket) => Ok(Some(socket)),
+            Err(_) => Ok(None),
+        },
+    }
```

Comment:
This looks good. Great work. The only problem is that it should return an error when all parsing fails instead of ignoring the value. We _do_ want to surface those errors to the user.